File: leetcode/medium/107-level-order-2/beautiful.py
"""
  @author Vortexx2
  Problem 102 - Binary Tree Level Order Traversal

  Runtime - 24 ms
  Memory Usage - 14.2 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
    res = []

    if not root:
      return res

    level = [root]
    while root and level:
      currNodes = []
      nextLevel = []

      for node in level:
        currNodes.append(node.val)
        if node.left:
          nextLevel.append(node.left)

        if node.right:
          nextLevel.append(node.right)

      res.append(currNodes)
      level = nextLevel

    return list(reversed(res))

File: leetcode/medium/107-level-order-2/beautiful.cpp
/**
 * @file beautiful.cpp
 * @author Vortexx2
 * @brief Problem 107 - Binary Tree Level Order Traversal II
 * @date 24-07-2021
 * 
 * Runtime - 4 ms
 * Memory Usage - 12.4 MB
 */
#include <iostream>
#include <algorithm>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<vector<int>> levelOrderBottom(TreeNode *root)
  {
    vector<vector<int>> res;
    if (!root)
      return res;

    vector<TreeNode *> level = {root};
    while (root && level.size())
    {
      vector<int> currNodes;
      vector<TreeNode*> nextLevel;

      for (auto &child: level) {
        currNodes.push_back(child->val);
        if (child->left) nextLevel.push_back(child->left);
        if (child->right) nextLevel.push_back(child->right);
      }

      res.push_back(currNodes);
      level = nextLevel;
    }
    reverse(res.begin(), res.end());
    return res;
  }
};